### PR TITLE
Better Liquid Evaporation, Space Cleaner Grenades actually work now, and more

### PR DESCRIPTION
**Pull request by MaxIsJoe**
CL: [Improvement] ReagentMixes have their creation date recorded now.
CL: [Improvement] Several optimizations have been done to how reagents are stored and spread on tiles. Expect some bugs and odd behaviors as we iron out the kinks in the future while keeping server performance as good as possible.
CL: [Improvement] Liquids evaporate faster now every (60 seconds + numbers of reagents inside a ReagentMix) in two frame intervals.
CL: [Improvement] Liquids that evaporate in high temperature areas generate water vapor gasses.
CL: [Improvement] Spilled Liquids that are more than 5 units can now be used to extinguish fires.
CL: [Improvement] Liquids will no longer spread through glass windows and airlocks. They can, however, still pass through grills.
CL: [Improvement] Liquids on the ground can now be removed using mops.
CL: [Improvement] Liquids that mix with other reagents on the floor will mix with their color.
CL: [Fix] Some pre-generated blood splats did not contain any blood reagents, this has been corrected.
CL: [Fix] Space Cleaner grenades were supposed to clean multiple tiles at once, this PR brings this intended behavior.

### CODE REVIEW

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs [hunk 1]
```diff
@@ -522,7 +579,7 @@ private void DistributeExcessToNearbyCells(ReagentMix excess, Vector3Int origin)
 				if (excess.Total <= 0) return;
 
 				// Skip if the neighbor is not passable
-				if (matrix.IsWallAt(neighbor, true)) continue;
+				if (matrix.IsWallAt(neighbor, true) || matrix.IsWindowAt(neighbor, true)) continue;
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: This will still go through doors though

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: It's intentional

Comment (corp-0) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: airlocks are sealed tho, they shouldn't let liquids pass

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: fair point

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: Updated to make it not pass through doors now.

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs [hunk 2]
```diff
@@ -522,7 +580,9 @@ private void DistributeExcessToNearbyCells(ReagentMix excess, Vector3Int origin)
 				if (excess.Total <= 0) return;
 
 				// Skip if the neighbor is not passable
-				if (matrix.IsWallAt(neighbor, true)) continue;
+				if (matrix.IsWallAt(neighbor, true)
+				    || matrix.IsWindowAt(neighbor, true)
+				    || matrix.GetNoGC(neighbor, true, new List<DoorMasterController>()).Any()) continue;
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: neighbor might have is Passable precalculated

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: public bool IsOccupied => Type == NodeType.Occupied;

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: tried that on the older PR, didn't work

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs [hunk 3]
```diff
@@ -506,35 +564,40 @@ public Tuple<bool, ReagentMix> HasReagentsAtTile(Vector3Int localPosInt)
...
-			foreach (var neighbor in neighbors)
+			foreach (var neighbor in currentCell.Neighbors)
 			{
 				if (excess.Total <= 0) return;
 
 				// Skip if the neighbor is not passable
-				if (matrix.IsWallAt(neighbor, true)) continue;
-				var neighborReagents = HasReagentsAtTile(neighbor);
+				if (neighbor.IsOccupied) continue;
+				var neighborReagents = HasReagentsAtTile(neighbor.LocalPosition);
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: you've already got the tile so you can just query the tiles for the reagent

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs [hunk 4]
```diff
@@ -506,35 +564,40 @@ public Tuple<bool, ReagentMix> HasReagentsAtTile(Vector3Int localPosInt)
...
 				// If the neighboring cell is empty, add the excess and break up the amount
 				if (neighborReagents.Item1 == false)
 				{
-					tileReagentMixes.Add(neighbor, excess.Split(excess.Total / 2));
-					CreateLiquidOverlay(neighbor, excess);
+					tileReagentMixes.Add(neighbor.LocalPosition, excess.Split(excess.Total / 2));
+					CreateLiquidOverlay(neighbor.LocalPosition, excess);
 #if UNITY_EDITOR
-					GameGizmomanager.AddNewLineStaticClient(null, currentCell.ToWorldInt(matrix), null, neighbor.ToWorldInt(matrix), Color.red);
+					GameGizmomanager.AddNewLineStaticClient(null,
```
Comment (Bod9001) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: maybe make this a setting on the  Meta layerdata like a bool on the inspector

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs (hunk 3 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: wdym

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs (hunk 4 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: it's under a pre-processor, it's fine if its left like this for the time being.

Comment (Bod9001) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaDataLayer.cs]: it's just having gizmos in the editor, it's an unnecessary performance impact  , would be better if it was a bool
 also means you're able to activate it in  the build if you VV it

**Review by Bod9001 — APPROVED** [review]

**Pull request merged**